### PR TITLE
enhance: use bump_schema_version to involve backfill and drop field compaction(#48808)

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: https://github.com/milvus-io/milvus/issues/48808

### CODE REVIEW

File: internal/datacoord/index_inspector.go [hunk 1]
```diff
@@ -158,50 +156,32 @@ func (i *indexInspector) createIndexesForSegment(ctx context.Context, segment *S
...
-	// physical backfill (e.g. AlterCollectionSchema adding a BM25 function): until backfill
-	// writes the new field's data, an index built on the empty binlog would silently return
-	// wrong results. For metadata-only schema changes (e.g. nullable AddCollectionField),
-	// the index builder handles null/default values, so no delay is needed.
-	// Within the physical-backfill branch we still double-check segment binlogs because a
-	// function output field that was backfilled in a previous schema version is already
-	// present in this segment.
-	coll := i.meta.GetCollection(segment.CollectionID)
-	collectionInPhysicalBackfill := coll != nil && coll.Schema != nil && coll.Schema.GetDoPhysicalBackfill()
 	var segmentBinlogFields map[int64]struct{}
```
Comment (liliu-z) [inline internal/datacoord/index_inspector.go:159]: When `bumpSchemaVersion.enabled=false`, the index inspector silently fails to build indexes on existing segments for fields added via `AddCollectionField`. This means newly added fields will have no index coverage on pre-existing data, leading to degraded query performance or incorrect search results with no user-visible error.

Comment (liliu-z) [inline internal/datacoord/index_inspector.go:159]: The `buildIndexCh` channel uses a `select { default: }` pattern that silently drops index-build notifications when the 1024-buffer channel is full. While the periodic indexInspector scan provides a fallback, dropped notifications cause latency spikes — affected segments won't get indexed until the next scan tick. This is not a permanent data loss but degrades index-build responsiveness.

File: configs/milvus.yaml [hunk 2]
```diff
@@ -772,7 +773,7 @@ dataCoord:
     clusteringCompactionUsage: 65535 # slot usage of clustering compaction task, setting it to 65536 means it takes up a whole worker.
     mixCompactionUsage: 4 # slot usage of mix compaction task.
     l0DeleteCompactionUsage: 8 # slot usage of l0 compaction task.
-    backfillCompactionUsage: 1 # slot usage of backfill compaction task.
+    bumpSchemaVersionCompactionUsage: 1 # slot usage of schema bump compaction task.
```
Comment (liliu-z) [inline configs/milvus.yaml:778]: The rename from `dataCoord.compaction.backfill.triggerInterval` to `dataCoord.compaction.bumpSchemaVersion.{enabled,triggerInterval}` and `dataCoord.slot.backfillCompactionUsage` to `dataCoord.slot.bumpSchemaVersionCompactionUsage` provides no backward-compatibility alias. Operators with explicit overrides for the old parameter names will silently fall back to defaults after upgrade, potentially causing unexpected behavior changes.

**Comment by liliu-z** [comment]
**internal/datacoord/compactor_common.go:333** In `compactionReadSchema`, when a StructArrayField has `len(childFields) == 0`, the entire StructArray is dropped from the read-side clone. The write side compensates via `GetAllFieldSchemas` + `GenerateEmptyArrayFromSchema`, so no data is permanently truncated. However, the synthesized-null array may produce a different on-disk column-group layout than segments with real child data. This edge case lacks test coverage.

File: internal/datacoord/index_inspector.go (hunk 1 above)
Comment (MrPresent-Han) [reply internal/datacoord/index_inspector.go:159]: I rechecked this path. With `bumpSchemaVersion.enabled=false`, `canCreateIndexForSegment` returns true immediately, so `indexInspector` still triggers index build for old segments.

For scalar fields added by `AddCollectionField`, the missing-field rows are handled inside the index-build pipeline rather than by `indexInspector`: DataCoord sends `LackBinlogRows = segment.NumRows - fieldBinlogRows`, and the C++ scalar index builders call `CacheRawDataAndFillMissing`, which fills those missing rows with the field default value or null before building the scalar index.

So this is not a regression for scalar added-field index coverage. The failure I found in the E2E was earlier in the pipeline: sort compaction failed before index build because the materialized empty nullable dense-vector Arrow array used `FixedSizeBinary` while the current writer path expects `Binary` for nullable dense vectors.

Comment (MrPresent-Han) [reply internal/datacoord/index_inspector.go:159]: This is a valid responsiveness concern, but it is not a correctness issue and it is not newly introduced by this PR.

`buildIndexCh` is a best-effort acceleration signal. If the channel is full, the signal can be dropped, but the segment/index metadata has already been persisted; `indexInspector` still periodically scans unindexed flushed segments via `getUnIndexTaskSegments()` and calls `createIndexesForSegment()`. So a dropped signal delays index build until the next inspector tick, but it does not leave the segment permanently unindexed.

This PR only adds a new producer for bump-schema-version compaction result segments and follows the existing non-blocking notification pattern used by flush, import, and mix compaction paths. I agree this path could be improved separately with a rated warning or metric when the channel is full.

File: configs/milvus.yaml (hunk 2 above)
Comment (MrPresent-Han) [reply configs/milvus.yaml:778]: I rechecked the release history for these config keys. The old `dataCoord.compaction.backfill.triggerInterval` and `dataCoord.slot.backfillCompactionUsage` keys were never released in a Milvus version, so there is no upgrade-compatibility requirement for existing user overrides.

Because no released deployment can already depend on those old keys, keeping only the new `bumpSchemaVersion` config names is safe and does not need backward-compatible aliases.

File: internal/datacoord/index_inspector.go [hunk 3]
```diff
@@ -213,18 +183,63 @@ func (i *indexInspector) createIndexesForSegment(ctx context.Context, segment *S
...
 			result[childFieldID] = struct{}{}
 		}
 	}
 	return result
 }
 
+func (i *indexInspector) canCreateIndexForSegment(ctx context.Context, segment *SegmentInfo, index *model.Index, segmentBinlogFields map[int64]struct{}) (bool, error) {
+	if !paramtable.Get().DataCoordCfg.BumpSchemaVersionCompactionEnabled.GetAsBool() {
+		return true, nil
+	}
```
Comment (tedxu) [inline internal/datacoord/index_inspector.go]: Concern: with `BumpSchemaVersionCompactionEnabled` defaulting to `false`, this short-circuit skips every schema-version check below. That means in default config the inspector can build an index on a function-output field whose binlog data does not yet exist on the segment. The deleted backfill code path had an equivalent safeguard (function output + missing binlog → defer). Once this short-circuit is taken we lose that protection too.

A cluster running default config that performs `AlterCollectionSchema` to add a function ends up worse off than before this PR: segments stay stale AND indexes get built on function-output fields without backfilled data, which is exactly the scenario the prior code guarded against.

Suggest either (a) flipping the default to on, or (b) keeping the function-output safeguard active independently of the flag.

File: internal/datacoord/index_inspector.go [hunk 4]
```diff
@@ -213,18 +183,63 @@ func (i *indexInspector) createIndexesForSegment(ctx context.Context, segment *S
...
 // getSegmentBinlogFields returns the set of field IDs that have binlog data in the segment.
-// binlog.GetFieldID() is a columnGroupID, not a real field ID; the actual field IDs
-// are always in binlog.GetChildFields().
+// StorageV2/V3 column groups report real field IDs through ChildFields; legacy entries may use FieldID directly.
 func getSegmentBinlogFields(segment *SegmentInfo) map[int64]struct{} {
 	result := make(map[int64]struct{})
 	for _, binlog := range segment.GetBinlogs() {
+		if len(binlog.GetChildFields()) == 0 {
+			result[binlog.GetFieldID()] = struct{}{}
+			continue
```
Comment (tedxu) [inline internal/datacoord/index_inspector.go:189]: The previous code explicitly warned that `binlog.FieldID` for storage v2 is a `columnGroupID`, not a real field id. Falling back to it when `ChildFields` is empty widens the matching surface: a V2 segment whose columnGroupID happens to equal a real `index.FieldID` will spuriously pass the `hasField` check in `canCreateIndexForSegment`. Worth gating this fallback on storage version (V1 only).

File: internal/datanode/compactor/bump_schema_version_compactor.go [hunk 5]
```diff
@@ -0,0 +1,1029 @@
...
+		totalRows += int64(wrapped.Len())
+		releaseWrappedRecord(wrapped, record)
+	}
+
+	if err := writer.Close(); err != nil {
+		return nil, err
+	}
+	writerClosed = true
+	// PK stats, BM25 stats, and expire quantiles come from BinlogRecordWriter.
+	insertLogs, _, _, manifestPath, expirQuantiles := writer.GetLogs()
```
Comment (tedxu) [inline internal/datanode/compactor/bump_schema_version_compactor.go]: PK statslog and BM25 statslog returned by `BinlogRecordWriter.GetLogs()` are discarded here. The resulting `CompactionSegment` sets neither `Field2StatslogPaths` nor `Bm25Logs`. Datacoord finalization (`internal/datacoord/meta.go` around `GetBm25Logs` / `internal/datacoord/task_stats.go`) reads these directly off the result segment.

So if a schema bump simultaneously drops a field AND the collection has a BM25 function, the new segment's metadata loses its BM25 statslogs. The manifest still carries them, but legacy/non-V3 consumers won't see them. `sort_compaction` preserves both. No test in the suite combines drop-field with BM25 — worth adding one and propagating the stats logs through here.

File: internal/datacoord/compaction_policy_bump_schema_version.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+// staleFlushedSegments returns schema-bump-eligible flushed segments for collectionID
+// whose SchemaVersion lags behind collectionSchemaVersion.
+func (policy *bumpSchemaVersionPolicy) staleFlushedSegments(collectionID int64, collectionSchemaVersion int32) []*chanPartSegments {
+	return GetSegmentsChanPart(policy.meta, collectionID, SegmentFilterFunc(func(segment *SegmentInfo) bool {
+		return isSchemaBumpDataSegment(segment) &&
+			!segment.isCompacting &&
+			segment.GetStorageVersion() >= storage.StorageV3 &&
+			segment.GetManifestPath() != "" &&
```
Comment (tedxu) [inline internal/datacoord/compaction_policy_bump_schema_version.go]: This hard `StorageVersion >= StorageV3 && ManifestPath != ""` gate is a semantic regression vs the deleted backfill policy (which had no such requirement). Any V2 segment with a stale schema becomes permanently invisible to this trigger — silently. No log, no metric, no operator-facing diagnostic. "Why didn't my BM25 backfill happen on my old segments?" will be hard to debug.

If this is intentional (schema evolution is V3-only by design), please at least emit a `log.RatedWarn` for skipped V2 segments and document the V3 prerequisite. Better: coordinate with `storageVersionUpgradePolicy` so V2 segments get promoted first, instead of orphaning them.

File: internal/datacoord/compaction_trigger_v2.go [hunk 7]
```diff
@@ -243,8 +240,8 @@ func (m *CompactionTriggerManager) loop(ctx context.Context) {
 			m.handleTicker(ctx, SingleTicker)
 		case <-storageVersionTicker.C:
 			m.handleTicker(ctx, StorageVersionTicker)
-		case <-backfillTicker.C:
-			m.handleTicker(ctx, BackfillTicker)
+		case <-bumpSchemaVersionTicker.C:
+			m.handleTicker(ctx, BumpSchemaVersionTicker)
```
Comment (tedxu) [inline internal/datacoord/compaction_trigger_v2.go:244]: Previously, schema bumps for the metadata-only case (no missing functions, no dropped fields) were applied inline via `TriggerInline` + `applyInlineView` — a fast in-memory meta update with no inspector slot.

This ticker now funnels every stale segment through a full compaction round-trip: allocate `PreAllocatedSegmentIDs` (2, not 1), enqueue on the inspector, run on a datanode worker, persist a compaction task. The compactor short-circuits via `runSchemaVersionBumpOnly` so it is cheap per task, but it is not free under inspector pressure — a collection with N stale segments now consumes N inspector slots vs. zero before for a pure metadata DDL like `AddField` with nullable default.

Consider preserving the inline fast path for the case where the segment needs neither function materialization nor field drop.

**Review by tedxu — COMMENTED** [review]
Inline notes on a handful of concerns. The biggest one is the default-off flag interacting badly with the removed safety gates in index_inspector.

File: internal/datacoord/index_inspector.go (hunk 3 above)
Comment (MrPresent-Han) [reply internal/datacoord/index_inspector.go]: Fixed by removing the feature-flag/schema-version early return. The function-output missing-binlog guard now runs unconditionally; normal fields are allowed without a physical binlog.

File: internal/datacoord/index_inspector.go (hunk 4 above)
Comment (MrPresent-Han) [reply internal/datacoord/index_inspector.go:189]: fixed

File: internal/datanode/compactor/bump_schema_version_compactor.go (hunk 5 above)
Comment (MrPresent-Han) [reply internal/datanode/compactor/bump_schema_version_compactor.go]: fixed

File: internal/datacoord/compaction_policy_bump_schema_version.go (hunk 6 above)
Comment (MrPresent-Han) [reply internal/datacoord/compaction_policy_bump_schema_version.go]: Addressed with minimal observability: stale segments that pass the schema-bump data filters but fail the V3+manifest prerequisite now emit a rate-limited warning with segment/schema/storage/manifest context before being skipped. Scheduling behavior stays unchanged.

File: internal/datacoord/compaction_trigger_v2.go (hunk 7 above)
Comment (MrPresent-Han) [reply internal/datacoord/compaction_trigger_v2.go:244]: DataCoord cannot safely decide whether the schema bump is metadata-only from its side alone. The decision depends on the segment's underlying V3 storage schema/manifest, which is inspected on the DataNode side to choose between full rewrite, partial backfill, or simply advancing the schema version metadata. Keeping this routed through the schema-bump compactor preserves that single decision point.

File: internal/datanode/compactor/bump_schema_version_compactor.go [hunk 8]
```diff
@@ -0,0 +1,1052 @@
...
+	defer reader.Close()
+
+	materializer, err := NewRecordMaterializer(t.plan.GetSchema(), t.plan.GetSchema().GetFunctions(), existingFields)
+	if err != nil {
+		return nil, err
+	}
+	defer materializer.Close()
+
+	alloc := allocator.NewLocalAllocator(t.plan.GetPreAllocatedLogIDs().GetBegin(), t.plan.GetPreAllocatedLogIDs().GetEnd())
+	// TODO(#50021): Support full TEXT LOB rewrite for schema-bump compaction.
```
Comment (liliu-z) [inline internal/datanode/compactor/bump_schema_version_compactor.go:436]: The full schema rewrite invokes `storage.NewBinlogRecordWriter` without `WithTextColumnConfigs` and without the LOB compaction context that mix/sort/clustering compaction sets up via `initLOBCompactionContext`. The inline `TODO(#50021): Support full TEXT LOB rewrite for schema-bump compaction.` acknowledges this gap. For a StorageV3 segment with a retained TEXT field backed by LOB Vortex files, a drop of another field will trigger this branch and produce a manifest whose TEXT column-group / `lob_file_refs` metadata isn't updated, breaking REWRITE_ALL/REUSE_ALL accounting and risking missing or stale LOB data once the old segment is GC'd. Either reject this branch when retained TEXT fields exist or wire the same LOB compaction context (`initLOBCompactionContext` + `WithTextColumnConfigs`) the other compactions use.
